ext/stackprof/stackprof.c

```diff
@@ -366,7 +366,7 @@ stackprof_record_sample_for_stack(int num)
 	    _stackprof.raw_samples = malloc(sizeof(VALUE) * _stackprof.raw_samples_capa);
 	}
 
-	if (_stackprof.raw_samples_capa <= _stackprof.raw_samples_len + num) {
+	while (_stackprof.raw_samples_capa <= _stackprof.raw_samples_len + (num + 2)) {
```

Comment:
Okay! I think this fixes it.

I believe this code was previously unsafe.

This is checking to see if the capacity of `raw_samples` is sufficient to cover `num` new entries, but it really needs to cover up to `num + 2` new entries, since each sample entry is

stack height, followed by n frame ids, followed by the sample count.

I *think* this was an issue even before this change, but I'm not able to reproduce a crash before this PR.

Regardless, I think this is safer now?

Reply:
I'm not sure I follow.. why would this need to be a loop?

Reply:
I see, because we just double every time.

Calling realloc multiple times seems silly. It should calculate the required size once, and then call realloc once.

Reply:
In practice, I suspect this will almost never need to call `realloc` more than once, because we set the initial buffer size to be 100 * the number of frames in the first sample, so for this to be called more than once, a given stack would have to be more than twice the size of all previous stacks combined. The `while` loop is just for safety, but I suspect in practice it'll never be a significant performance burden.

Reply:
Okay fair enough.